winpr/libwinpr/file/generic.c

```diff
@@ -0,0 +1,613 @@
+	}
+	CopyMemory(pFileSearch->lpPattern, &lpFileName[index + 1], length);
+	pFileSearch->lpPattern[length] = '\0';
+
+	/* Check if the path is a directory */
+
+	if (lstat(pFileSearch->lpPath, &fileStat) < 0)
+	{
+		FindClose(pFileSearch);
+		return INVALID_HANDLE_VALUE; /* stat error */
+	}
+
+	if (S_ISDIR(fileStat.st_mode) == 0)
+	{
+		FindClose(pFileSearch);
+		return INVALID_HANDLE_VALUE; /* not a directory */
+	}
+
+	/* Open directory for reading */
+	pFileSearch->pDir = opendir(pFileSearch->lpPath);
+
+	if (!pFileSearch->pDir)
+	{
+		FindClose(pFileSearch);
+		return INVALID_HANDLE_VALUE; /* failed to open directory */
+	}
+
+	while ((pFileSearch->pDirent = readdir(pFileSearch->pDir)) != NULL)
+	{
+		if ((strcmp(pFileSearch->pDirent->d_name, ".") == 0) || (strcmp(pFileSearch->pDirent->d_name, "..") == 0))
+		{
+			/* skip "." and ".." */
+			continue;
+		}
+
+		if (FilePatternMatchA(pFileSearch->pDirent->d_name, pFileSearch->lpPattern))
+		{
+			strcpy(lpFindFileData->cFileName, pFileSearch->pDirent->d_name);
+			return (HANDLE) pFileSearch;
+		}
+	}
+
+	FindClose(pFileSearch);
+	return INVALID_HANDLE_VALUE;
+}
+
+HANDLE FindFirstFileW(LPCWSTR lpFileName, LPWIN32_FIND_DATAW lpFindFileData)
+{
+	return NULL;
+}
+
+HANDLE FindFirstFileExA(LPCSTR lpFileName, FINDEX_INFO_LEVELS fInfoLevelId, LPVOID lpFindFileData,
+						FINDEX_SEARCH_OPS fSearchOp, LPVOID lpSearchFilter, DWORD dwAdditionalFlags)
+{
+	return NULL;
+}
+
+HANDLE FindFirstFileExW(LPCWSTR lpFileName, FINDEX_INFO_LEVELS fInfoLevelId, LPVOID lpFindFileData,
+						FINDEX_SEARCH_OPS fSearchOp, LPVOID lpSearchFilter, DWORD dwAdditionalFlags)
+{
+	return NULL;
+}
+
+BOOL FindNextFileA(HANDLE hFindFile, LPWIN32_FIND_DATAA lpFindFileData)
+{
+	WIN32_FILE_SEARCH* pFileSearch;
+
+	if (!hFindFile)
+		return FALSE;
+
+	if (hFindFile == INVALID_HANDLE_VALUE)
+		return FALSE;
+
+	pFileSearch = (WIN32_FILE_SEARCH*) hFindFile;
+
+	while ((pFileSearch->pDirent = readdir(pFileSearch->pDir)) != NULL)
+	{
+		if (FilePatternMatchA(pFileSearch->pDirent->d_name, pFileSearch->lpPattern))
+		{
+			strcpy(lpFindFileData->cFileName, pFileSearch->pDirent->d_name);
+			return TRUE;
+		}
+	}
+
+	return FALSE;
+}
+
+BOOL FindNextFileW(HANDLE hFindFile, LPWIN32_FIND_DATAW lpFindFileData)
+{
+	return FALSE;
+}
+
+BOOL FindClose(HANDLE hFindFile)
+{
+	WIN32_FILE_SEARCH* pFileSearch;
+	pFileSearch = (WIN32_FILE_SEARCH*) hFindFile;
+
+	if (pFileSearch)
+	{
+		free(pFileSearch->lpPath);
+		free(pFileSearch->lpPattern);
+
+		if (pFileSearch->pDir)
+			closedir(pFileSearch->pDir);
+
+		free(pFileSearch);
+		return TRUE;
+	}
+
+	return FALSE;
+}
+
+BOOL CreateDirectoryA(LPCSTR lpPathName, LPSECURITY_ATTRIBUTES lpSecurityAttributes)
+{
+	if (!mkdir(lpPathName, S_IRUSR | S_IWUSR | S_IXUSR))
+		return TRUE;
+
+	return FALSE;
+}
+
+BOOL CreateDirectoryW(LPCWSTR lpPathName, LPSECURITY_ATTRIBUTES lpSecurityAttributes)
+{
+	return FALSE;
+}
+
+BOOL RemoveDirectoryA(LPCSTR lpPathName)
+{
+	if (!rmdir(lpPathName))
```

Comment:
@hardening this was just split moved out from the old file.c I didn't review the code. But your suggestion makes sense.